multi: Implement DCP0011 PoW hash consensus vote.

davecgh opened the pull request:
This implements the agenda for voting on changing the hash function to BLAKE3, resetting the target difficulty, and changing the difficulty algorithm to ASERT (Absolutely Scheduled Exponentially weighted Rising Targets) as defined in DCP0011 along with consensus tests.

In terms of the overall effects, this includes updates to:

- The validation logic for the proof of work hash and required block difficulty
- Enforcement of block submission via the `getwork` and `submitblock` RPCs
- Mining template generation
- The output of the `getwork` and `notifywork` RPCs
- The internal CPU miner

Also note that this does not implement the block version bump that will ultimately be needed by the mining code since there are multiple consensus votes gated behind it and will therefore be done separately.

The following are some graphs that show the behavior and responsiveness of the difficulty algorithm for select data in the included reference tests.  In particular, the tests which use main network parameters and random solve times per a Poisson distribution for a stable hashrate, rapidly increasing hashrate, and rapidly decreasing hashrate. The first graph shows the per-block difficulty for the three scenarios all combined.  The next three graphs show the individual histograms of the solve times overlaid with the normalized Poisson distribution to show the test data for these tests follows the expected distribution as claimed:

![decred_mainnet_asert_difficulty_algorithm_select_test_data](https://github.com/decred/dcrd/assets/2115102/5f573007-2c96-4705-80ef-2f91ff432f0b)

![decred_mainnet_asert_difficulty_algorithm_select_test_data_solve_time_histogram_stable](https://github.com/decred/dcrd/assets/2115102/de2f2f17-2600-4fe5-93bd-9d607065b8f2)

![decred_mainnet_asert_difficulty_algorithm_select_test_data_solve_time_histogram_increasing](https://github.com/decred/dcrd/assets/2115102/5d457118-dc18-4d9d-b424-c973a51f4cef)

![decred_mainnet_asert_difficulty_algorithm_select_test_data_solve_time_histogram_decreasing](https://github.com/decred/dcrd/assets/2115102/c8cfbf95-e3e3-45c8-817e-f9ab04abe35d)


The various changes have been separated into a series of commits to help make the review process easier.

The following is an overview of the changes:

- Separate block hash and proof of work hash into distinct concepts
- Add new `PowHashV1` and `PowHashV2` methods to `wire` for computing the existing BLAKE256 and new BLAKE3 hashes, respectively
- Add BLAKE3 support to the test chain generator
- Consolidate the logic that serializes the data that represents work to be solved in the `getwork` RPC and `notifywork` `WebSocket` notification
- Add new method to `blockchain/standalone` named `CheckProofOfWorkHash` to independently verify that a provided hash is less than a provided compact target difficulty without also additionally checking that the compact difficulty is within a valid range
- Introduce convenience function for determining if the vote passed and is now active
- Introduce convenience function for determining whether or not the agenda is forced active on networks other than the main network
- Add new method to `blockchain/standalone` for calculating a target difficulty using the new ASERT difficulty algorithm
- Add per-network parameters to `chaincfg` for the new difficulty algorithm including the initial starting difficulty for BLAKE3 and the half life
- Add ASERT support to the test chain generator
- Modify block validation to enforce BLAKE3 as the proof of work hash algorithm in accordance with the state of the vote
- Modify block validation to enforce target difficulties per the ASERT algorithm based on a reset initial starting difficulty in accordance with the state of the vote
- Update internal CPU miner to solve blocks with either BLAKE256 or BLAKE3 in accordance with the state of the vote
- Update the `getwork`, `notifywork`, and `submitblock` RPCs to support BLAKE3 in accordance with the state of the vote
- Add tests for `CheckProofOfWorkHash`
- Add tests for determining if the agenda is active for both mainnet and testnet
- Add comprehensive tests for the ASERT difficulty calculation in `blockchain/standalone`
- Add tests for the `getwork` RPC including submission of a block solved with BLAKE3
- Add tests to ensure proper behavior for the hash function and difficulty semantics before and after the vote agenda is active
- Updates `blockchain/standalone` documentation

davecgh commented on blockchain/standalone/pow_test.go:
```diff
@@ -353,3 +393,1613 @@ func TestCheckProofOfWork(t *testing.T) {
+			{17, 2157, 0x1c016a22},
+			{18, 2160, 0x1c014388},
+		},
+	}}
```
This should add an additional test to trigger potential internal calculation overflow boundaries in case the arbitrary precision integers are changed out for a fixed precision type in the future.

Also, adding a couple of tests using simnet parameters would be ideal too since the proof of work limit on simnet is 255 bits which means the internal calculations will need more than 256 bits without resorting to contrived values as the aforementioned suggested overflow boundary test would.

davecgh replied:
Both of these have been addressed and are now part of the relevant commit.

davecgh commented on blockchain/standalone/pow.go:
```diff
@@ -218,3 +218,169 @@ func CheckProofOfWork(powHash *chainhash.Hash, difficultyBits uint32, powLimit *
+	// Calculate the target difficulty by multiplying the provided starting
+	// target difficulty by an exponential scaling factor that is determined
+	// based on how far ahead or behind the ideal schedule the given time delta
+	// is along with a half life that acts smoothing factor.
```
Done.

davecgh commented on blockchain/standalone/pow.go:
```diff
@@ -218,3 +218,169 @@ func CheckProofOfWork(powHash *chainhash.Hash, difficultyBits uint32, powLimit *
+	//   nextDiff = min(max(nextDiff, 1), powLimit)
+	//
+	// NOTE: The division by the half life uses Quo instead of Div because it
+	// must be use truncated division (which is truncated towards zero and Quo
```
Updated.

davecgh commented on blockchain/standalone/pow_test.go:
```diff
@@ -353,3 +393,1613 @@ func TestCheckProofOfWork(t *testing.T) {
+		halfLifeSecs       int64
+	}
+
+	// mockMainNetParams hosues the mainnet parameters to use throughout the
```
Corrected.

davecgh commented on blockchain/standalone/pow_test.go:
```diff
@@ -353,3 +393,1613 @@ func TestCheckProofOfWork(t *testing.T) {
+		halfLifeSecs:       43200,
+	}
+
+	// mockTestNetParams hosues the mainnet parameters to use throughout the
```
Corrected.

davecgh commented on blockchain/standalone/pow.go:
```diff
@@ -218,3 +218,180 @@ func CheckProofOfWork(powHash *chainhash.Hash, difficultyBits uint32, powLimit *
+	//   n = ---- = x >> 16
+	//       2^16
+	//
+	//   f = x (mod 16) = x & 0xffff
```
```suggestion
f = x (mod 2^16) = x & 0xffff
```

davecgh replied:
Updated.

davecgh commented on blockchain/standalone/pow.go:
```diff
@@ -218,3 +218,180 @@ func CheckProofOfWork(powHash *chainhash.Hash, difficultyBits uint32, powLimit *
+	//                       2^16
+	//
+	// Note that by exponent rules 2^n / 2^16 = 2^(n - 16).  This takes
+	// advantage of that property to reduce that multiplication by 2^n and
```
Changed.

davecgh commented on blockchain/standalone/pow.go:
```diff
@@ -218,3 +218,180 @@ func CheckProofOfWork(powHash *chainhash.Hash, difficultyBits uint32, powLimit *
+	// are used for this implementation, it doesn't make any difference from a
+	// correctness standpoint, however, it does potentially lower the amount of
+	// memory for the arbitrary precision type and can be used to help prevent
+	// overflow in implementations that used fixed precision types.
```
Done.

davecgh commented:
I added some additional tests to `TestBlake3PowSemantics` that exercise the new difficulty algorithm selection semantics as well.

matheusd reviewed (commented):
First pass at reviewing, still looking at the two most important commits.

davecgh commented on wire/go.mod:
```diff
@@ -5,6 +5,10 @@ go 1.17
 require (
 	github.com/davecgh/go-spew v1.1.1
 	github.com/decred/dcrd/chaincfg/chainhash v1.0.2
+	lukechampine.com/blake3 v1.1.7
```
Nice.  I see it was just tagged last week and I hadn't checked again.  I went ahead and reviewed the changes between 1.7.1 and 1.2.1 and agree that none of them are relevant in terms of the blake3 hashing itself.  I did notice that v1.2.0 bumped the `go` directive in the `go.mod` to version 1.17 though which does bring module graph pruning and lazy loading support, so that is a nice benefit.

Updated to v1.2.1.

davecgh commented on internal/blockchain/blockindex.go:
```diff
@@ -625,14 +633,26 @@ func (bi *blockIndex) HaveBlock(hash *chainhash.Hash) bool {
+	// order to start zeroing out the bits used here.  Not only is that
+	// impossible for all intents and purposes, the only effect would be
+	// reducing overall memory savings due to increased collisions among the
+	// shortened keys.
```
Yeah, I noticed this as well, but I figured it wasn't an issue because it doesn't gain anything (worst case is a bit more memory usage, but overall it'd still be quite a bit less than it would be without the short keys) and, as you noted, there is a (slight) disincentive in the associated mining time.

I also thought about tying it to the pow hash instead of the block hash, but that would either (a) require storing the pow hash along with every `blockNode` thereby entirely defeating any memory savings the short keys provide (actually slightly worse versus not doing the short keys at all due to overhead); or (b) calculating the pow hash on every lookup which is significantly more expensive in terms of CPU usage and node lookups happen super frequently and thus would overall be way worse in all cases.

davecgh commented on internal/blockchain/agendas_test.go:
```diff
@@ -1017,6 +1017,117 @@ func TestSubsidySplitDeployment(t *testing.T) {
 
+// testBlake3PowDeployment ensures the deployment of the blake3 proof of work
+// agenda activates for the provided network parameters.
+func testBlake3PowDeployment(t *testing.T, params *chaincfg.Params) {
```
This should also include logic to ensure the blake3 anchor block selection works as intended since I noticed there are not any tests currently hitting that path.

davecgh replied:
Added.

davecgh commented on blockchain/standalone/pow.go:
```diff
@@ -218,3 +218,180 @@ func CheckProofOfWork(powHash *chainhash.Hash, difficultyBits uint32, powLimit *
+	// guaranteed to be positive and a maximum of 17 bits, so it is safe to cast
+	// to a uint32.
+	const (
+		polyCoeff1 uint64 = 195766423245049 // 0.695502049 * 2^48
```
Notice that double checking the math in the comment gives `ceil(0.695502049 * 2^48) = 195766423044489`  which is not `195766423245049`.  This is because it was calculated based on a more precise version of the polynomial than the one in the comments.

This comment should be updated to the more precise value of `ceil(0.695502049712533 * 2^48)` to avoid confusion.

davecgh replied:
Updated.

davecgh commented on blockchain/standalone/pow.go:
```diff
@@ -218,3 +218,180 @@ func CheckProofOfWork(powHash *chainhash.Hash, difficultyBits uint32, powLimit *
+	// to a uint32.
+	const (
+		polyCoeff1 uint64 = 195766423245049 // 0.695502049 * 2^48
+		polyCoeff2 uint64 = 971821376       // 0.2262698 * 2^32
```
Similarly, this should be updated to the more precise value of `ceil(0.2262697964 * 2^32)`.

davecgh replied:
Updated.

davecgh commented on blockchain/standalone/pow.go:
```diff
@@ -218,3 +218,180 @@ func CheckProofOfWork(powHash *chainhash.Hash, difficultyBits uint32, powLimit *
+	//
+	// In particular, the goal cubic polynomial approximation is:
+	//
+	//   2^x ~= 1 + 0.695502049x + 0.2262698x^2 + 0.0782318x^3 for 0 <= x < 1
```
Done.

matheusd commented:
commit "chaingen: Add ASERT difficulty algorithm support. " has 

> This adds support for generating blocks **the** require the new difficulty

davecgh commented on blockchain/standalone/pow.go:
```diff
@@ -182,22 +182,217 @@ func CheckProofOfWorkRange(difficultyBits uint32, powLimit *big.Int) error {
+
+	// Decompose the exponent into integer and fractional parts.  Since the
+	// exponent is using 64.16 fixed point, the bottom 16 bits are the
+	// fractional part and the integer part is the exponent right shifted by 16.
```
Since the exponent is signed, this should explicitly say it is **arithmetic** right shifted because a logical right shift will not produce the correct result for negative values.

davecgh replied:
> commit "chaingen: Add ASERT difficulty algorithm support. " has
> 
> > This adds support for generating blocks **the** require the new difficulty

Fixed.

davecgh replied:
Added.

davecgh commented:
> commit "chaingen: Add ASERT difficulty algorithm support. " has
> 
> > This adds support for generating blocks **the** require the new difficulty

Fixed.

davecgh commented on blockchain/standalone/pow.go:
```diff
@@ -218,3 +218,181 @@ func CheckProofOfWork(powHash *chainhash.Hash, difficultyBits uint32, powLimit *
+	//
+	// Per DCP0011, the goal equation is:
+	//
+	//   nextDiff = min(max(startDiff * 2^((Δt - Δh*Ib)/halfLife), 1), powLimit)
```
Yes, the change to per-block difficulty is very intentional.  I'm glad to see you that. The DCP (which isn't up yet) calls this out more explicitly (though it still needs a bit more refinement) along with the motivation behind it.

In short though, your observation is correct. It improves the responsiveness and allows the algorithm to better maintain the target block time (aka ideal block schedule). Another important aspect is that it means the network can more easily adjust to large swings in the hash power, notable large drops, versus the current EMA approach with a larger retarget interval since it only needs a single block at the no-longer-ideal difficulty versus potentially an entire interval.

davecgh commented on blockchain/standalone/pow.go:
```diff
@@ -218,3 +218,181 @@ func CheckProofOfWork(powHash *chainhash.Hash, difficultyBits uint32, powLimit *
+	// NOTE: The division by the half life uses Quo instead of Div because it
+	// must be truncated division (which is truncated towards zero as Quo
+	// implements) as opposed to the Euclidean division that Div implements.
+	idealTimeDelta := heightDelta * targetSecsPerBlock
+	exponentBig := big.NewInt(timeDelta - idealTimeDelta)
+	exponentBig.Lsh(exponentBig, 16)
+	exponentBig.Quo(exponentBig, big.NewInt(halfLife))
```
Yes, given the stated assumption, and for the main network, that's correct about 24 bits.

In fact, one of the reference test data sets I created involves generating a series of blocks starting from the hardest difficulty that are each spaced at the half life so that it halves the difficulty each block until it gets to the easiest difficulty.  The final time delta is 9744000 which results in a difference from the ideal schedule of 9676800 which is indeed 24 bits.

davecgh commented on blockchain/standalone/pow.go:
```diff
@@ -218,3 +218,181 @@ func CheckProofOfWork(powHash *chainhash.Hash, difficultyBits uint32, powLimit *
+	// Note that a full unsigned 64-bit type is required to avoid overflow in
+	// the internal 16.48 fixed point calculation.  Also, the overall result is
+	// guaranteed to be positive and a maximum of 17 bits, so it is safe to cast
+	// to a uint32.
```
Was all of this obvious enough from the comment, or do you think it would be beneficial to expand a bit?

davecgh commented on internal/blockchain/chain.go:
```diff
@@ -271,6 +273,24 @@ type BlockChain struct {
+	// cachedBlake3WorkDiffAnchor houses a cached anchor point to use for the
+	// difficulty algorithm defined in DCP0011.
+	//
+	// It is only set when the blake3 proof of work agenda has determined to be
```
Added.

matheusd commented on blockchain/standalone/pow_test.go:
```diff
@@ -393,3 +393,1676 @@ func TestCheckProofOfWork(t *testing.T) {
+			{8, 2400, 0x1d00ffff},
+			{9, 2700, 0x1d00ffff},
+			{10, 3000, 0x1d00ffff},
+		},
```
This asserts CalcASERT maintains the

matheusd commented on blockchain/standalone/pow_test.go:
```diff
@@ -393,3 +393,1676 @@ func TestCheckProofOfWork(t *testing.T) {
+			{9223372036854775809, 2147484247, 0x1b00a5a6},
+		},
+	}, {
+		name: "mainnet: blocks across potential internal calculation overflow" +
+			" boundaries",
```
While debugging the tests, I noticed there was no test exercising the case where the internal `frac64` is set to its maximum amount (`0xfff`). Here is a test case to exercise that:

```go
{                                                                                                 
        name:          "mainnet: blocks that exercise max fractional part in internal calculation",  
        params:        &mockMainNetParams,                                                           
        startDiffBits: 0x1b00a5a6,                                                                   
        startHeight:   0,                                                                            
        startTime:     0,                                                                            
        perInstanceTests: []perInstanceTest{                                                         
                {1, 299, 0x1b00a5a5},                                                                
                {2, 599, 0x1b00a5a5},                                   
                {3, 899, 0x1b00a5a5},                                                                
                {4, 1199, 0x1b00a5a5},                                                               
        },                                                                                           
}
```

davecgh replied:
Nice.  It's a good idea to add a vector for this case as well.  I will not however that these overflow tests do hit the edges where the inner addition requires all 64 bits for a correct result (all of them require the full 64 bits for the inner 16.48 addition).

Thanks for the additional test vectors!

Added.

davecgh replied:
Nice.  It's a good idea to add a vector for this case as well.  I will not however that these overflow tests do hit the edges where the inner addition requires all 64 bits for a correct result (all of them require the full 64 bits for the inner 16.48 addition).

Thanks for the additional test vectors!

Added.

matheusd commented on internal/blockchain/difficulty.go:
```diff
@@ -212,20 +211,152 @@ func (b *BlockChain) calcNextRequiredDifficulty(prevNode *blockNode, newBlockTim
+	// the next call will involve a descendant of this anchor as opposed to some
+	// other anchor on an entirely unrelated side chain.
+	if anchor != nil {
+		b.cachedBlake3WorkDiffAnchor.Store(anchor)
```
This is the only place (other than at startup when the agenda is forced) where this is set. 

There could be some flickering of this value, right after the activation block, if multiple sidechains end up being mined, but this should be resolved fairly quickly (and in any case this func will return very fast, because the first iteration of the loop will already succeed).

This is also the reason why the access to `cachedBlake3WorkDiffAnchor` at the top of this function is gated by checking whether `anchor.isAncestor(prevNode)`.

matheusd commented on blockchain/standalone/pow_test.go:
```diff
@@ -393,3 +393,1676 @@ func TestCheckProofOfWork(t *testing.T) {
+			{11, 10, 0x1e185487},
+			{12, 12, 0x1e1b4e81},
+		},
+	}}
```
This is missing a test when both `heightDelta` and `timeDelta` == 0

davecgh replied:
Good call.  Added to `"mainnet: steady 300s blocks at difficulty 0x1b00a5a6"`

matheusd commented on internal/blockchain/difficulty.go:
```diff
@@ -212,20 +211,152 @@ func (b *BlockChain) calcNextRequiredDifficulty(prevNode *blockNode, newBlockTim
+	//
+	// However, since the initial difficulty is reset with the agenda, no
+	// additional offsets are needed.
+	timeDelta := prevNode.timestamp - blake3Anchor.timestamp
+	heightDelta := prevNode.height - blake3Anchor.height
```
When `blake3Anchor == prevNode` (i.e. validating the first block where this is activated), `timeDelta == 0` and `heightDelta == 0`.

davecgh replied:
Correct, and this ensures the required difficulty for the first block when the agenda activates is exactly the initial starting difficulty.

davecgh commented on blockchain/standalone/pow_test.go:
```diff
@@ -353,3 +393,1676 @@ func TestCheckProofOfWork(t *testing.T) {
+		startHeight:   0,
+		startTime:     0,
+		perInstanceTests: []perInstanceTest{
+			{1, 43200 * 16, 0x1d00ffff},
```
These should have an additional comment describing what they intent to hit because it's probably not obvious otherwise.

davecgh replied:
Added.

davecgh commented on internal/blockchain/difficulty.go:
```diff
@@ -212,20 +211,152 @@ func (b *BlockChain) calcNextRequiredDifficulty(prevNode *blockNode, newBlockTim
+	rcai := int64(b.chainParams.RuleChangeActivationInterval)
+	svh := b.chainParams.StakeValidationHeight
+	finalNodeHeight := calcWantHeight(svh, rcai, prevNode.height+1)
+	candidate := prevNode.Ancestor(finalNodeHeight)
```
This is theoretically true (and part of why the following checks help protect against misuse nil checks), but, in practice, due to the precondition that the function is required to be called with the agenda active, it can't ever be the genesis block or `nil`.

This is because either:
(a) the anchor was the genesis block already and thus the function would've returned it above; or
(b) the agenda must have gone through activation which required an interval for it to be locked in meaning the floor here would be the final node of the locked in interval.

davecgh commented on internal/blockchain/difficulty.go:
```diff
@@ -212,20 +211,152 @@ func (b *BlockChain) calcNextRequiredDifficulty(prevNode *blockNode, newBlockTim
+	svh := b.chainParams.StakeValidationHeight
+	finalNodeHeight := calcWantHeight(svh, rcai, prevNode.height+1)
+	candidate := prevNode.Ancestor(finalNodeHeight)
+	for candidate != nil && candidate.parent != nil {
```
Correct.  Similar to the above discussion, the agenda is either always active in which case it is enforced starting from block 1 or it must go through the activation process which requires stake validation height + a voting interval + a locked in interval.

davecgh replied:
After further consideration, I don't think using the genesis block as the anchor for networks where the agenda is always active is the best choice.

I say this because the timestamp in the genesis has to be set to something when the code creates it, but by the time the code is merged and released, a long period of time might have passed.  That means that blocks based on it as the anchor will heavily be behind schedule leading to a major difficulty drop for a ton of initial blocks.

That doesn't matter too much for the simulation network where blocks being the easiest difficulty is the norm anyway, but it could very well be an issue for future versions of the test network.

Instead, I think networks where the agenda is always active should be special cases so that the first block is required to be the initial starting difficulty and then all subsequent blocks should use the first block as the anchor.

davecgh replied:
I pushed a new commit (that will be merged after review) to implement the aforementioned behavior of special casing networks where the agenda is always active so that the first block is required to be the initial starting difficulty and then all subsequent blocks should use the first block as the anchor.

davecgh replied:
The commit has been squashed.

davecgh commented on blockchain/standalone/pow_test.go:
```diff
@@ -393,3 +393,1676 @@ func TestCheckProofOfWork(t *testing.T) {
+			{268, 782911, 0x1d00c69d},
+			{269, 785122, 0x1d00ccca},
+			{270, 792634, 0x1d00e5e6},
+			{271, 806060, 0x1d00ffff},
```
This was for a 90% decrease.  That said, looking at this again based on your comment, I'm going to update this one to start at a higher initial difficulty in order to avoid clipping at the proof of work limit so quickly which should provide test data that is more likely to discover issues in other implementations.

davecgh commented:
I've updated the reference test data for the rapidly decreasing and rapidly increasing hash rates and updated the difficulty graph in the PR description accordingly.

In particular, the rapidly decreasing test data now targets 1000s average solve times and starts at a higher initial difficulty in order to avoid clipping at the proof of work limit so quickly which should provide test data that is more likely to discover issues in other implementations.

Similarly, the rapidly increasing test data now targets 15s average solve times.

davecgh commented on blockchain/chaingen/generator.go:
```diff
@@ -1021,6 +1082,309 @@ func (g *Generator) CalcNextRequiredDifficulty() uint32 {
+// will panic.
+//
+// This function is safe for concurrent access.
+func calcASERTDiff(startDiffBits uint32, powLimit *big.Int, targetSecsPerBlock,
```
`chaingen` is intended to be entirely independent code from the code that consensus uses since otherwise it would be the equivalent of testing if something is equal to itself and thus wouldn't be good at catching regressions.  You'll notice that `chaingen` doesn't use `standalone` nor `blockchain` (back when it was exported) for anything else either for the same reason.

Also, ultimately the `standalone.CalcASERTDiff` will likely be optimized to avoid using arbitrary precision integers and changed over to use a fixed precision type like `math/uint256` at which point having the separate implementation still based on a big ints will be even more valuable.

jrick commented on blockchain/chaingen/generator.go:
```diff
@@ -1021,6 +1082,309 @@ func (g *Generator) CalcNextRequiredDifficulty() uint32 {
+	return compact
+}
+
+// CalcASERTDiff calculates an absolutely scheduled exponentially weighted
```
since we're keeping this, should lowercase the func name in this comment.

davecgh commented on internal/blockchain/difficulty.go:
```diff
@@ -212,20 +211,152 @@ func (b *BlockChain) calcNextRequiredDifficulty(prevNode *blockNode, newBlockTim
+	// Note that it will always be set to the genesis block on networks where
+	// the agenda is always active.
+	anchor := b.cachedBlake3WorkDiffAnchor.Load()
+	if anchor != nil && anchor.IsAncestorOf(prevNode) {
```
This has now changed due to the move to using the first block as the anchor once it has been mined.  It is now only ever set in this very function.

jrick commented on internal/blockchain/validate.go:
```diff
@@ -1014,6 +1035,160 @@ func (b *BlockChain) isOldBlockVersionByMajority(header *wire.BlockHeader, block
+		blake3Diff := b.calcNextBlake3Diff(prevNode)
+		if header.Bits != blake3Diff {
+			str := fmt.Sprintf("block difficulty of %d is not the expected "+
+				"value of %d", header.Bits, blake3Diff)
```
we may want to change all of the formatted errors in this function to describe which hashing algorithm was used to improve debuggability.

jrick replied:
sounds good to me

davecgh replied:
Updated.

davecgh commented on internal/blockchain/validate.go:
```diff
@@ -1014,6 +1035,163 @@ func (b *BlockChain) isOldBlockVersionByMajority(header *wire.BlockHeader, block
+	if !isMainNet(b.chainParams) {
+		minBlake3BlockVersion++
+	}
+	isBlake3PossiblyActive := uint32(header.Version) >= minBlake3BlockVersion
```
Looking through this again, I see this could also check for a height > svh+rcai*2 here now since a new check was added above for the always active case meaning the only remaining possible way for the agenda to be active for it to have been voted in which requires voting to be possible (svh), at least one interval of voting (rcai), and one interval of locked in (rcai) to have passed.

davecgh replied:
I've updated this accordingly.

jrick commented on internal/rpcserver/rpcserver.go:
```diff
@@ -3849,21 +3875,50 @@ func handleGetWorkRequest(ctx context.Context, s *Server) (interface{}, error) {
 }
 
+// minInt is a helper function to return the minimum of two ints.  This avoids
+// the need to cast to floats.
```
not sure what floats have to do with this?

davecgh replied:
Both of the standard lib [math.Min](https://pkg.go.dev/math#Min) and [math.max](https://pkg.go.dev/math#Max) take floats.

davecgh commented on blockchain/standalone/pow_test.go:
```diff
@@ -353,3 +393,1700 @@ func TestCheckProofOfWork(t *testing.T) {
+		startHeight      int64  // starting height for calculation
+		startTime        int64  // starting time for calculation
+		perInstanceTests []perInstanceTest
+	}{{
```
Since these are all reference test vectors from the DCP, I'm going to split them out in a separate JSON-formatted test vectors file and then update this code to parse that JSON and run the tests.

Having the test vectors external makes it much easier for other implementations to consume them and also helps ensure continuity between the DCP and these tests.

davecgh replied:
Done.

davecgh commented:
As mentioned in [this comment](https://github.com/decred/dcrd/pull/3115#discussion_r1213536777), the reference test vectors have been split out into a separate JSON-formatted file.  Please keep in mind when doing a final review pass.

jrick commented:
I'm guessing that the json was generated from existing go data structures.

Can we keep those generators, perhaps in the testdata directory, so they can be rerun to make changes in the future?

davecgh commented:
> I'm guessing that the json was generated from existing go data structures.
> 
> Can we keep those generators, perhaps in the testdata directory, so they can be rerun to make changes in the future?

Yeah, I generated it from the original test data, but then I replaced the tests to just read in them JSON and run it instead of having the exact same data in two places.

jrick commented:
I'm just thinking ahead to if/when we will want to make changes to the test data, it may be easier/cleaner to do so by editing existing go source with the structures, as opposed to hand editing the json.  At the very least, it removes the possibility of not following the same whitespace as existing data.

davecgh commented:
> I'm just thinking ahead to if/when we will want to make changes to the test data, it may be easier/cleaner to do so by editing existing go source with the structures, as opposed to hand editing the json. At the very least, it removes the possibility of not following the same whitespace as existing data.

For reference, we discussed this a bit more offline and I decided to go ahead and treat the JSON as the source of truth since it's pretty simple to edit it as needed given it is fully keyed unlike array-based JSON that is much harder.

matheusd commented on internal/blockchain/difficulty.go:
```diff
@@ -212,20 +211,186 @@ func (b *BlockChain) calcNextRequiredDifficulty(prevNode *blockNode, newBlockTim
+//
+// This function MUST only be called with the blake3 proof of work agenda active
+// after having gone through a vote.  That is to say it MUST NOT be called when
+// the agenda is forced to always be active.
```
Rechecked this condition in call sites after squash.

matheusd commented on internal/blockchain/difficulty.go:
```diff
@@ -212,20 +211,186 @@ func (b *BlockChain) calcNextRequiredDifficulty(prevNode *blockNode, newBlockTim
+//
+// This function MUST only be called with the blake3 proof of work agenda active
+// and, unless the agenda is always forced to be active, with the chain state
+// lock held (for writes).
```
Rechecked this condition in call sites after squash.

matheusd commented on blockchain/go.mod:
```diff
@@ -20,6 +20,10 @@ require (
 
 replace github.com/decred/dcrd/chaincfg/v3 => ../chaincfg
+
+replace github.com/decred/dcrd/wire => ../wire
```
This replacement directive needs to be moved to commit "multi: Separate block hash and proof of work hash", otherwise `run_tests.sh` fails in the intervening commits.

davecgh replied:
Nice catch.  I didn't notice because of my local workspace.  I also noticed the next commit needed to be tidied due to the addition of the new dep to `wire` as well.  All updated accordingly.

matheusd reviewed (commented):
Made a second review after the changes to the networks where the blake3 agenda is force-activated.

Looks like only another couple of nits to fix.

matheusd commented:
Commit "chaingen: Add ASERT difficulty algorithm support" doesn't successfully run `run_tests.sh` due to the changes introduced in the next commit to the `chaincfg` package

jrick commented on internal/blockchain/validate.go:
```diff
@@ -1062,7 +1238,7 @@ func (b *BlockChain) checkBlockHeaderPositional(header *wire.BlockHeader, prevNo
 		blockHeight := prevNode.height + 1
 		if b.minTestNetTarget != nil &&
-			expDiff <= standalone.BigToCompact(b.minTestNetTarget) &&
+			header.Bits <= standalone.BigToCompact(b.minTestNetTarget) &&
```
i believe this could use b.minTestNetDiffBits instead of doing the BigToCompact calculation again.

davecgh commented:
> Commit "chaingen: Add ASERT difficulty algorithm support" doesn't successfully run `run_tests.sh` due to the changes introduced in the next commit to the `chaincfg` package

Good catch since I prefer for every intermediate commit to continue to build and pass tests.  I've moved the relevant changes to add the new parameters to a separate commit (`chaincfg: Add params for DCP0011 blake3 diff calc.`) prior to that commit in order to resolve it.

matheusd reviewed (approved):
Tests passing on all commits now!
